// @vnedyalk0v/react19-simple-maps@2.0.2/dist/utils.js
import { feature, mesh } from 'topojson-client';
import { cache } from 'react';
import { prefetchDNS, preconnect, preload } from 'react-dom';

// Helper functions to create branded types
const createLongitude = (value) => value;
const createLatitude = (value) => value;
const createCoordinates = (lon, lat) => [
    createLongitude(lon),
    createLatitude(lat),
];

/**
 * Calculates coordinates from zoom transform
 * @param w - Width of the map
 * @param h - Height of the map
 * @param t - Zoom transform object
 * @returns Branded coordinates
 */
function getCoords(w, h, t) {
    const xOffset = (w * t.k - w) / 2;
    const yOffset = (h * t.k - h) / 2;
    const lon = w / 2 - (xOffset + t.x) / t.k;
    const lat = h / 2 - (yOffset + t.y) / t.k;
    return createCoordinates(lon, lat);
}

/**
 * Creates a standardized geography fetch error
 * @param type - Error type
 * @param message - Error message
 * @param url - URL that caused the error (optional)
 * @param originalError - Original error that caused this error (optional)
 * @returns GeographyError instance
 */
function createGeographyFetchError(type, message, url, originalError) {
    const error = new Error(message);
    error.name = 'GeographyError';
    error.type = type;
    error.timestamp = new Date().toISOString();
    if (url) {
        error.geography = url;
    }
    if (originalError) {
        error.cause = originalError;
        if (originalError.stack) {
            error.stack = originalError.stack;
        }
        error.details = {
            originalMessage: originalError.message,
            originalName: originalError.name,
        };
    }
    return error;
}

/**
 * Simple URL validation to avoid circular dependency
 * @param url - URL to validate
 * @returns Validated URL string
 */
function validateURL(url) {
    if (!url || typeof url !== 'string') {
        throw new Error('URL must be a non-empty string');
    }
    // Basic URL validation
    try {
        new URL(url);
        return url.trim();
    }
    catch {
        throw new Error('Invalid URL format');
    }
}
const DEFAULT_GEOGRAPHY_FETCH_CONFIG = {
    TIMEOUT_MS: 10000, // 10 seconds
    MAX_RESPONSE_SIZE: 50 * 1024 * 1024, // 50MB
    ALLOWED_CONTENT_TYPES: ['application/json', 'application/geo+json'],
    ALLOWED_PROTOCOLS: ['https:'], // HTTPS only by default
    ALLOW_HTTP_LOCALHOST: false, // Disabled by default for security
    STRICT_HTTPS_ONLY: true, // Strict HTTPS-only mode by default
};
// Development configuration (can be enabled explicitly)
const DEVELOPMENT_GEOGRAPHY_FETCH_CONFIG = {
    ...DEFAULT_GEOGRAPHY_FETCH_CONFIG,
    ALLOWED_PROTOCOLS: ['https:', 'http:'],
    ALLOW_HTTP_LOCALHOST: true, // Allow HTTP for localhost in development
    STRICT_HTTPS_ONLY: false,
};
// Current active configuration (defaults to secure)
let GEOGRAPHY_FETCH_CONFIG = DEFAULT_GEOGRAPHY_FETCH_CONFIG;
/**
 * Configure geography fetching security settings
 * @param config - Security configuration to apply
 */
function configureGeographySecurity(config) {
    GEOGRAPHY_FETCH_CONFIG = {
        ...DEFAULT_GEOGRAPHY_FETCH_CONFIG,
        ...config,
    };
}
/**
 * Enable development mode with relaxed security (use with caution)
 * @param allowHttpLocalhost - Whether to allow HTTP for localhost
 */
function enableDevelopmentMode(allowHttpLocalhost = true) {
    if (process.env.NODE_ENV === 'production') {
        // eslint-disable-next-line no-console
        console.warn('Attempted to enable development mode in production - ignoring for security');
        return;
    }
    GEOGRAPHY_FETCH_CONFIG = {
        ...DEVELOPMENT_GEOGRAPHY_FETCH_CONFIG,
        ALLOW_HTTP_LOCALHOST: allowHttpLocalhost,
    };
    // eslint-disable-next-line no-console
    console.warn('Development mode enabled with relaxed security settings. Do not use in production!');
}
/**
 * Checks if a hostname is a private IP address
 * @param hostname - The hostname to check
 * @returns True if the hostname is a private IP address
 */
function isPrivateIPAddress(hostname) {
    // Skip non-IP hostnames
    if (!hostname || hostname === 'localhost') {
        return false;
    }
    // IPv4 private ranges
    const ipv4PrivateRanges = [
        /^10\./, // 10.0.0.0/8
        /^172\.(1[6-9]|2[0-9]|3[01])\./, // 172.16.0.0/12
        /^192\.168\./, // 192.168.0.0/16
        /^127\./, // 127.0.0.0/8 (loopback)
        /^169\.254\./, // 169.254.0.0/16 (link-local)
    ];
    // Check IPv4 private ranges
    for (const range of ipv4PrivateRanges) {
        if (range.test(hostname)) {
            return true;
        }
    }
    // IPv6 private ranges (simplified check)
    const ipv6PrivateRanges = [
        /^::1$/, // ::1 (loopback)
        /^fe80:/, // fe80::/10 (link-local)
        /^fc00:/, // fc00::/7 (unique local)
        /^fd00:/, // fd00::/8 (unique local)
    ];
    // Check IPv6 private ranges
    for (const range of ipv6PrivateRanges) {
        if (range.test(hostname)) {
            return true;
        }
    }
    return false;
}
/**
 * Validates a geography URL for security and format compliance
 * @param url - The URL to validate
 * @throws {Error} If the URL is invalid or insecure
 */
function validateGeographyUrl(url) {
    // Use comprehensive URL validation from input-validation module
    const validatedUrl = validateURL(url);
    try {
        const parsedUrl = new URL(validatedUrl);
        // Strict HTTPS-only mode
        if (GEOGRAPHY_FETCH_CONFIG.STRICT_HTTPS_ONLY) {
            if (parsedUrl.protocol !== 'https:') {
                throw createGeographyFetchError('SECURITY_ERROR', `Strict HTTPS-only mode: ${parsedUrl.protocol} is not allowed. Only HTTPS is permitted.`, url);
            }
        }
        else {
            // Check protocol security with configured protocols
            if (!GEOGRAPHY_FETCH_CONFIG.ALLOWED_PROTOCOLS.includes(parsedUrl.protocol)) {
                const allowedProtocols = GEOGRAPHY_FETCH_CONFIG.ALLOWED_PROTOCOLS.join(', ');
                throw createGeographyFetchError('SECURITY_ERROR', `Unsupported protocol: ${parsedUrl.protocol}. Only ${allowedProtocols} are allowed.`, url);
            }
            // HTTP protocol validation
            if (parsedUrl.protocol === 'http:') {
                // Check if HTTP localhost is explicitly allowed
                if (!GEOGRAPHY_FETCH_CONFIG.ALLOW_HTTP_LOCALHOST) {
                    throw createGeographyFetchError('SECURITY_ERROR', 'HTTP protocol is disabled for security. Use HTTPS or enable development mode explicitly.', url);
                }
                // If HTTP localhost is allowed, validate hostname
                if (parsedUrl.hostname !== 'localhost' &&
                    parsedUrl.hostname !== '127.0.0.1') {
                    throw createGeographyFetchError('SECURITY_ERROR', 'HTTP protocol is only allowed for localhost. Use HTTPS for remote URLs.', url);
                }
                // Additional production check
                if (process.env.NODE_ENV === 'production') {
                    throw createGeographyFetchError('SECURITY_ERROR', 'HTTP localhost access is not allowed in production', url);
                }
                // Development warning for HTTP localhost usage
                // eslint-disable-next-line no-console
                console.warn(`Security Warning: Using HTTP for localhost (${url}). This should only be used in development.`);
            }
        }
        // Additional security checks for localhost access
        if (parsedUrl.hostname === 'localhost' ||
            parsedUrl.hostname === '127.0.0.1') {
            if (process.env.NODE_ENV === 'production') {
                throw createGeographyFetchError('SECURITY_ERROR', 'Localhost access is not allowed in production', url);
            }
        }
        // Validate against private IP ranges (additional security)
        if (isPrivateIPAddress(parsedUrl.hostname)) {
            throw createGeographyFetchError('SECURITY_ERROR', `Access to private IP address ${parsedUrl.hostname} is not allowed`, url);
        }
    }
    catch (error) {
        if (error instanceof TypeError) {
            throw createGeographyFetchError('VALIDATION_ERROR', `Invalid URL format: ${url}`, url, error);
        }
        throw error;
    }
}
/**
 * Validates response content type
 * @param response - The fetch response to validate
 * @throws {Error} If content type is invalid
 */
function validateContentType(response) {
    const contentType = response.headers.get('content-type');
    if (!contentType) {
        throw createGeographyFetchError('VALIDATION_ERROR', 'Missing Content-Type header');
    }
    const isValidType = GEOGRAPHY_FETCH_CONFIG.ALLOWED_CONTENT_TYPES.some((type) => contentType.toLowerCase().includes(type));
    if (!isValidType) {
        throw createGeographyFetchError('VALIDATION_ERROR', `Invalid content type: ${contentType}. Expected one of: ${GEOGRAPHY_FETCH_CONFIG.ALLOWED_CONTENT_TYPES.join(', ')}`);
    }
}
/**
 * Fast pre-check of Content-Length header to reject obviously oversized responses.
 * NOTE: This is only a pre-check — the header can be omitted or falsified.
 * Use {@link readResponseWithSizeLimit} for authoritative enforcement.
 * @param response - The fetch response to validate
 * @throws {Error} If Content-Length exceeds the configured maximum
 */
async function validateResponseSize(response) {
    const contentLength = response.headers.get('content-length');
    if (contentLength) {
        const size = parseInt(contentLength, 10);
        if (size > GEOGRAPHY_FETCH_CONFIG.MAX_RESPONSE_SIZE) {
            throw createGeographyFetchError('VALIDATION_ERROR', `Response too large: ${size} bytes. Maximum allowed: ${GEOGRAPHY_FETCH_CONFIG.MAX_RESPONSE_SIZE} bytes`);
        }
    }
}
/**
 * Reads the response body as an ArrayBuffer while enforcing a hard byte-count limit.
 * Protects against responses that omit or falsify Content-Length.
 * @param response - The fetch response to read
 * @param maxBytes - Maximum allowed bytes (defaults to GEOGRAPHY_FETCH_CONFIG.MAX_RESPONSE_SIZE)
 * @returns The response body as ArrayBuffer
 * @throws {Error} If the body exceeds the byte limit
 */
async function readResponseWithSizeLimit(response, maxBytes = GEOGRAPHY_FETCH_CONFIG.MAX_RESPONSE_SIZE) {
    const reader = response.body?.getReader();
    // Fallback: if ReadableStream is not available, read the whole body and check size.
    // NOTE: response.arrayBuffer() loads the entire response into memory before the size
    // check runs, which can cause high memory usage or OOM for very large responses.
    // The ReadableStream path above is preferred as it streams data and enforces the
    // size limit incrementally, failing fast without buffering the full payload.
    if (!reader) {
        const buffer = await response.arrayBuffer();
        if (buffer.byteLength > maxBytes) {
            throw createGeographyFetchError('VALIDATION_ERROR', `Response too large: ${buffer.byteLength} bytes exceeds limit of ${maxBytes} bytes`);
        }
        return buffer;
    }
    const chunks = [];
    let totalBytes = 0;
    for (;;) {
        const { done, value } = await reader.read();
        if (done)
            break;
        totalBytes += value.byteLength;
        if (totalBytes > maxBytes) {
            reader.cancel().catch(() => { });
            throw createGeographyFetchError('VALIDATION_ERROR', `Response too large: exceeded limit of ${maxBytes} bytes`);
        }
        chunks.push(value);
    }
    // Concatenate chunks into a single ArrayBuffer
    const result = new Uint8Array(totalBytes);
    let offset = 0;
    for (const chunk of chunks) {
        result.set(chunk, offset);
        offset += chunk.byteLength;
    }
    return result.buffer;
}
/**
 * Validates that the parsed data is a valid geography object
 * @param data - The parsed JSON data to validate
 * @throws {Error} If data is not a valid geography object
 */
function validateGeographyData(data) {
    if (!data || typeof data !== 'object') {
        throw createGeographyFetchError('VALIDATION_ERROR', 'Invalid geography data: not a valid object');
    }
    const obj = data;
    if (!obj.type ||
        (obj.type !== 'Topology' && obj.type !== 'FeatureCollection')) {
        throw createGeographyFetchError('VALIDATION_ERROR', `Invalid geography data: expected Topology or FeatureCollection, got ${obj.type}`);
    }
}

/**
 * Known SRI hashes for common geography data sources
 * These hashes are automatically generated and verified
 * Run 'npm run generate-sri' to update these hashes
 */
const KNOWN_GEOGRAPHY_SRI = {
    // World Atlas from unpkg.com - Countries data
    'https://unpkg.com/world-atlas@2/countries-110m.json': {
        algorithm: 'sha384',
        hash: 'sha384-yOCJ+8ShBm8UDqtAVtAvxTDDf4gXo5edxl/YG0FmVC5OTmqVLl7utuVGBDEeZWHf',
        enforceIntegrity: true,
    },
    'https://unpkg.com/world-atlas@2/countries-50m.json': {
        algorithm: 'sha384',
        hash: 'sha384-Aw4s9pX1PTPntIYkZ/qV9IYiF5Gv8eTl6Dd/TT56zfO1Wwd+owFwYUuuXNUMrWkc',
        enforceIntegrity: true,
    },
    // World Atlas from unpkg.com - Land data
    'https://unpkg.com/world-atlas@2/land-110m.json': {
        algorithm: 'sha384',
        hash: 'sha384-5oFOGoMd0tkagYW08lVco4uAi7XDEDBwBxOdeKx+SA1ihbsHiR/aFAJGretluTzG',
        enforceIntegrity: true,
    },
    'https://unpkg.com/world-atlas@2/land-50m.json': {
        algorithm: 'sha384',
        hash: 'sha384-c0VeCJd1wVbV5WQZNjf1hcMqPr9QXweEArnbdgS1k75TBNjta2M/NddyAulA/Glb',
        enforceIntegrity: true,
    },
};
const DEFAULT_SRI_CONFIG = {
    enforceForKnownSources: true,
    enforceForAllSources: false, // Don't enforce for all sources by default
    allowUnknownSources: true, // Allow unknown sources by default
    customSRIMap: {},
};
let currentSRIConfig = DEFAULT_SRI_CONFIG;
/**
 * Configure SRI enforcement settings
 * @param config - SRI enforcement configuration
 */
function configureSRI(config) {
    currentSRIConfig = {
        ...DEFAULT_SRI_CONFIG,
        ...config,
    };
}
/**
 * Enable strict SRI mode (enforce for all sources)
 */
function enableStrictSRI() {
    currentSRIConfig = {
        ...currentSRIConfig,
        enforceForKnownSources: true,
        enforceForAllSources: true,
        allowUnknownSources: false,
    };
}
/**
 * Disable SRI enforcement (not recommended for production)
 */
function disableSRI() {
    if (process.env.NODE_ENV === 'production') {
        // eslint-disable-next-line no-console
        console.warn('Disabling SRI in production is not recommended for security');
    }
    currentSRIConfig = {
        ...currentSRIConfig,
        enforceForKnownSources: false,
        enforceForAllSources: false,
        allowUnknownSources: true,
    };
}
/**
 * Calculate SHA hash of data
 * @param data - Data to hash
 * @param algorithm - Hash algorithm to use
 * @returns Promise resolving to base64-encoded hash
 */
async function calculateHash(data, algorithm) {
    // Use Web Crypto API (available in browsers and Node.js 16+)
    const hashBuffer = await globalThis.crypto.subtle.digest(algorithm, data);
    const hashArray = new Uint8Array(hashBuffer);
    // Convert to base64 (browser-compatible)
    let hashBase64;
    if (typeof globalThis.btoa !== 'undefined') {
        // Browser environment
        hashBase64 = globalThis.btoa(String.fromCharCode(...hashArray));
    }
    else {
        // Node.js environment - use manual base64 encoding
        const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/';
        let result = '';
        let i = 0;
        while (i < hashArray.length) {
            const a = hashArray[i++] || 0;
            const b = i < hashArray.length ? hashArray[i++] || 0 : 0;
            const c = i < hashArray.length ? hashArray[i++] || 0 : 0;
            const bitmap = (a << 16) | (b << 8) | c;
            result += chars.charAt((bitmap >> 18) & 63);
            result += chars.charAt((bitmap >> 12) & 63);
            result +=
                i - 2 < hashArray.length ? chars.charAt((bitmap >> 6) & 63) : '=';
            result += i - 1 < hashArray.length ? chars.charAt(bitmap & 63) : '=';
        }
        hashBase64 = result;
    }
    return hashBase64;
}
/**
 * Validate response integrity using SRI
 * @param response - Fetch response to validate
 * @param url - URL of the resource
 * @param expectedSRI - Expected SRI configuration
 * @returns Promise resolving to validated response
 */
async function validateSRI(response, url, expectedSRI) {
    // Clone response to avoid consuming the body
    const responseClone = response.clone();
    const data = await responseClone.arrayBuffer();
    // Validate using the ArrayBuffer approach
    await validateSRIFromArrayBuffer(data, url, expectedSRI);
    return response;
}
/**
 * Validate ArrayBuffer integrity using SRI
 * @param arrayBuffer - Data to validate
 * @param url - URL of the resource
 * @param expectedSRI - Expected SRI configuration
 * @returns Promise that resolves if validation passes, throws if it fails
 */
async function validateSRIFromArrayBuffer(arrayBuffer, url, expectedSRI) {
    // Calculate hash based on algorithm
    const algorithmMap = {
        sha256: 'SHA-256',
        sha384: 'SHA-384',
        sha512: 'SHA-512',
    };
    const calculatedHash = await calculateHash(arrayBuffer, algorithmMap[expectedSRI.algorithm]);
    const expectedHash = expectedSRI.hash.replace(`${expectedSRI.algorithm}-`, '');
    if (calculatedHash !== expectedHash) {
        const sriError = new Error(`Subresource Integrity check failed for ${url}. Expected ${expectedSRI.algorithm}-${expectedHash}, got ${expectedSRI.algorithm}-${calculatedHash}`);
        sriError.expectedHash = expectedSRI.hash;
        sriError.calculatedHash = `${expectedSRI.algorithm}-${calculatedHash}`;
        sriError.algorithm = expectedSRI.algorithm;
        throw createGeographyFetchError('SECURITY_ERROR', sriError.message, url, sriError);
    }
}
/**
 * Check if SRI validation is required for a URL
 * @param url - URL to check
 * @returns SRI configuration if validation is required, null otherwise
 */
function getSRIForUrl(url) {
    // Check custom SRI map first
    if (currentSRIConfig.customSRIMap[url]) {
        return currentSRIConfig.customSRIMap[url];
    }
    // Check known sources
    if (KNOWN_GEOGRAPHY_SRI[url] && currentSRIConfig.enforceForKnownSources) {
        return KNOWN_GEOGRAPHY_SRI[url];
    }
    // If enforcing for all sources but no SRI available
    if (currentSRIConfig.enforceForAllSources) {
        if (!currentSRIConfig.allowUnknownSources) {
            throw createGeographyFetchError('SECURITY_ERROR', `SRI enforcement is enabled but no integrity hash is available for ${url}`, url);
        }
    }
    return null;
}
/**
 * Add custom SRI configuration for a URL
 * @param url - URL to add SRI for
 * @param sri - SRI configuration
 */
function addCustomSRI(url, sri) {
    currentSRIConfig.customSRIMap[url] = sri;
}
/**
 * Generate SRI hash for a given URL (utility for developers)
 * This function fetches the resource and calculates its hash
 * @param url - URL to generate SRI for
 * @param algorithm - Hash algorithm to use
 * @returns Promise resolving to SRI hash string
 */
async function generateSRIHash(url, algorithm = 'sha384') {
    try {
        const response = await fetch(url);
        if (!response.ok) {
            throw new Error(`Failed to fetch ${url}: ${response.statusText}`);
        }
        const data = await response.arrayBuffer();
        const algorithmMap = {
            sha256: 'SHA-256',
            sha384: 'SHA-384',
            sha512: 'SHA-512',
        };
        const hash = await calculateHash(data, algorithmMap[algorithm]);
        return `${algorithm}-${hash}`;
    }
    catch (error) {
        throw createGeographyFetchError('GEOGRAPHY_LOAD_ERROR', `Failed to generate SRI hash for ${url}: ${error instanceof Error ? error.message : 'Unknown error'}`, url, error instanceof Error ? error : new Error(String(error)));
    }
}
/**
 * Validate multiple URLs and generate SRI configuration
 * Utility function for setting up SRI for multiple geography sources
 * @param urls - Array of URLs to generate SRI for
 * @param algorithm - Hash algorithm to use
 * @returns Promise resolving to SRI configuration map
 */
async function generateSRIForUrls(urls, algorithm = 'sha384') {
    const sriMap = {};
    for (const url of urls) {
        try {
            const hash = await generateSRIHash(url, algorithm);
            sriMap[url] = {
                algorithm,
                hash,
                enforceIntegrity: true,
            };
        }
        catch (error) {
            // eslint-disable-next-line no-console
            console.warn(`Failed to generate SRI for ${url}:`, error);
        }
    }
    return sriMap;
}

/** Maximum number of redirect hops allowed */
const MAX_REDIRECTS = 5;
/**
 * Creates fetch options with security headers and timeout.
 * Uses `redirect: 'manual'` so each redirect hop can be validated against the URL policy.
 * @param signal - AbortController signal for timeout
 * @returns Fetch options object
 */
function createSecureFetchOptions(signal) {
    return {
        signal,
        headers: {
            Accept: GEOGRAPHY_FETCH_CONFIG.ALLOWED_CONTENT_TYPES.join(', '),
            'Cache-Control': 'public, max-age=3600', // Cache for 1 hour
        },
        // Security headers
        mode: 'cors',
        credentials: 'omit', // Don't send credentials
        redirect: 'manual', // Handle redirects manually to validate each hop
    };
}
/**
 * Follows redirects manually, validating each hop against the URL security policy.
 * Prevents redirect-based SSRF bypasses.
 * @param url - The initial URL to fetch
 * @param options - Fetch options (must have redirect: 'manual')
 * @returns The final non-redirect response
 */
async function fetchWithRedirectValidation(url, options) {
    let currentUrl = url;
    for (let hop = 0; hop < MAX_REDIRECTS; hop++) {
        const response = await fetch(currentUrl, options);
        // Not a redirect — return directly
        if (response.status < 300 || response.status >= 400) {
            return response;
        }
        // Consume the redirect response body to release connection resources
        try {
            await response.arrayBuffer();
        }
        catch {
            // Ignore body-consumption errors — they must not mask redirect handling
        }
        // Extract and validate the redirect target
        const location = response.headers.get('location');
        if (!location) {
            throw createGeographyFetchError('SECURITY_ERROR', `Redirect response (HTTP ${response.status}) missing Location header`, currentUrl);
        }
        // Resolve relative redirects against current URL
        const redirectUrl = new URL(location, currentUrl).href;
        // Validate the redirect target against the same URL security policy
        validateGeographyUrl(redirectUrl);
        currentUrl = redirectUrl;
    }
    throw createGeographyFetchError('SECURITY_ERROR', `Too many redirects (exceeded ${MAX_REDIRECTS} hops)`, url);
}
/**
 * Creates an abort controller with timeout
 * @param timeoutMs - Timeout in milliseconds
 * @returns Object with controller and cleanup function
 */
function createTimeoutController(timeoutMs) {
    const controller = new AbortController();
    const timeoutId = setTimeout(() => {
        controller.abort();
    }, timeoutMs);
    return {
        controller,
        cleanup: () => clearTimeout(timeoutId),
    };
}
/**
 * Handles fetch errors and converts them to geography-specific errors
 * @param error - The original error
 * @param url - The URL that was being fetched
 * @returns A GeographyError
 */
function handleFetchError(error, url) {
    if (error instanceof Error) {
        if (error.name === 'AbortError') {
            return createGeographyFetchError('GEOGRAPHY_LOAD_ERROR', `Request timeout after ${GEOGRAPHY_FETCH_CONFIG.TIMEOUT_MS}ms`, url, error);
        }
        if (error.name === 'TypeError' && error.message.includes('fetch')) {
            return createGeographyFetchError('GEOGRAPHY_LOAD_ERROR', `Network error: Unable to fetch geography from ${url}`, url, error);
        }
        if (error.message.includes('Invalid geography data')) {
            return createGeographyFetchError('GEOGRAPHY_PARSE_ERROR', error.message, url, error);
        }
    }
    // Re-throw if it's already a GeographyError
    if (error instanceof Error && 'type' in error) {
        return error;
    }
    // Default error
    return createGeographyFetchError('GEOGRAPHY_LOAD_ERROR', error instanceof Error ? error.message : 'Unknown error occurred', url, error instanceof Error ? error : undefined);
}
/**
 * Parses JSON from ArrayBuffer with proper error handling
 * @param arrayBuffer - The response data as ArrayBuffer
 * @param url - The URL for error context
 * @returns Parsed geography data
 */
async function parseGeographyFromArrayBuffer(arrayBuffer, url) {
    try {
        const text = new TextDecoder().decode(arrayBuffer);
        const data = JSON.parse(text);
        validateGeographyData(data);
        return data;
    }
    catch (jsonError) {
        if (jsonError instanceof SyntaxError) {
            throw createGeographyFetchError('GEOGRAPHY_PARSE_ERROR', 'Invalid JSON format in geography data', url, jsonError);
        }
        throw jsonError;
    }
}
/**
 * Fetch geography data with full security validation.
 *
 * @deprecated Since v2.1.0 — use {@link fetchGeographiesCache} instead for
 * cached, secure fetching. This function now delegates to the hardened pipeline
 * but swallows errors for backward compatibility.
 *
 * @param url - The URL to fetch geography data from
 * @returns Promise resolving to geography data or undefined on error
 */
async function fetchGeographies(url) {
    if (typeof process !== 'undefined' &&
        process?.env?.NODE_ENV !== 'production') {
        // eslint-disable-next-line no-console
        console.warn('fetchGeographies is deprecated. Use fetchGeographiesCache for secure, cached fetching.');
    }
    try {
        return await fetchGeographiesCache(url);
    }
    catch {
        return undefined;
    }
}
/**
 * Secure, cached geography fetching with comprehensive validation
 * This function is cached using React's cache() for optimal performance
 */
const fetchGeographiesCache = cache(async (url) => {
    // Validate URL before making request
    validateGeographyUrl(url);
    // Check if SRI validation is required
    const sriConfig = getSRIForUrl(url);
    // Create timeout controller
    const { controller, cleanup } = createTimeoutController(GEOGRAPHY_FETCH_CONFIG.TIMEOUT_MS);
    try {
        // Make secure fetch request with redirect validation
        const response = await fetchWithRedirectValidation(url, createSecureFetchOptions(controller.signal));
        cleanup();
        // Validate response
        if (!response.ok) {
            throw createGeographyFetchError('GEOGRAPHY_LOAD_ERROR', `HTTP ${response.status}: ${response.statusText}`, url);
        }
        // Validate content type and fast pre-check of Content-Length
        validateContentType(response);
        await validateResponseSize(response);
        // Read body with hard streaming size limit (guards against falsified Content-Length)
        const arrayBuffer = await readResponseWithSizeLimit(response);
        // Handle SRI validation if required
        if (sriConfig) {
            await validateSRIFromArrayBuffer(arrayBuffer, url, sriConfig);
        }
        // Parse JSON from the already-read ArrayBuffer
        return await parseGeographyFromArrayBuffer(arrayBuffer, url);
    }
    catch (error) {
        cleanup();
        throw handleFetchError(error, url);
    }
});
/**
 * Preloads geography data for better performance
 * @param url - The URL to preload
 */
function preloadGeography$1(url) {
    // Import and use the preload utility with immediate flag
    Promise.resolve().then(function () { return preloading; })
        .then(({ preloadGeography: preloadUtil }) => {
        preloadUtil(url, true); // immediate = true
    })
        .catch(() => {
        // Silently handle import errors
    });
    // Also preload the actual data
    fetchGeographiesCache(url).catch(() => {
        // Silently ignore preload errors
    });
}

/**
 * Checks if the input is a string (URL)
 * @param geo - Geography data or URL
 * @returns True if input is a string
 */
function isString(geo) {
    return typeof geo === 'string';
}
/**
 * Extracts features from topology data
 * @param topology - Topology object
 * @param parseGeographies - Optional parser function
 * @returns Array of features
 */
function extractFeaturesFromTopology(topology, parseGeographies) {
    const objectKeys = Object.keys(topology.objects);
    if (objectKeys.length === 0) {
        return [];
    }
    // Get the first object (usually countries, states, etc.)
    const firstObjectKey = objectKeys[0];
    if (!firstObjectKey) {
        return [];
    }
    const geometryObject = topology.objects[firstObjectKey];
    if (!geometryObject) {
        return [];
    }
    const featureCollection = feature(topology, geometryObject);
    const features = 'features' in featureCollection ? featureCollection.features || [] : [];
    return parseGeographies ? parseGeographies(features) : features;
}
/**
 * Extracts features from FeatureCollection
 * @param featureCollection - FeatureCollection object
 * @param parseGeographies - Optional parser function
 * @returns Array of features
 */
function extractFeaturesFromCollection(featureCollection, parseGeographies) {
    const features = featureCollection.features || [];
    return parseGeographies ? parseGeographies(features) : features;
}
/**
 * Extracts features from various geography data formats
 * @param geographies - Geography data (Topology, FeatureCollection, or Feature array)
 * @param parseGeographies - Optional parser function for features
 * @returns Array of features
 */
function getFeatures(geographies, parseGeographies) {
    // Handle array of features
    if (Array.isArray(geographies)) {
        return parseGeographies ? parseGeographies(geographies) : geographies;
    }
    // Handle Topology
    if (geographies.type === 'Topology') {
        return extractFeaturesFromTopology(geographies, parseGeographies);
    }
    // Handle FeatureCollection
    if (geographies.type === 'FeatureCollection') {
        return extractFeaturesFromCollection(geographies, parseGeographies);
    }
    return [];
}
/**
 * Extracts mesh data from topology for borders and outlines
 * @param topology - Topology object
 * @returns Mesh data with outline and borders, or null if not available
 */
function extractMeshFromTopology(topology) {
    const objectKeys = Object.keys(topology.objects);
    if (objectKeys.length === 0) {
        return null;
    }
    const firstObjectKey = objectKeys[0];
    if (!firstObjectKey) {
        return null;
    }
    const geometryObject = topology.objects[firstObjectKey];
    if (!geometryObject) {
        return null;
    }
    try {
        // Generate outline (exterior boundaries)
        const outline = mesh(topology, geometryObject, (a, b) => a === b);
        // Generate borders (interior boundaries)
        const borders = mesh(topology, geometryObject, (a, b) => a !== b);
        return { outline, borders };
    }
    catch {
        return null;
    }
}
/**
 * Extracts mesh data from geography data
 * @param geographies - Geography data (only Topology supports mesh)
 * @returns Mesh data or null
 */
function getMesh(geographies) {
    // Only Topology supports mesh generation
    if (geographies &&
        typeof geographies === 'object' &&
        !Array.isArray(geographies) &&
        'type' in geographies &&
        geographies.type === 'Topology') {
        return extractMeshFromTopology(geographies);
    }
    return null;
}
/**
 * Prepares mesh data by generating SVG paths
 * @param outline - Outline geometry
 * @param borders - Borders geometry
 * @param path - D3 path generator
 * @returns Object with SVG path strings
 */
function prepareMesh(outline, borders, path) {
    const result = {};
    if (outline) {
        const outlinePath = path(outline);
        if (outlinePath) {
            result.outline = outlinePath;
        }
    }
    if (borders) {
        const bordersPath = path(borders);
        if (bordersPath) {
            result.borders = bordersPath;
        }
    }
    return result;
}
/**
 * Prepares features by generating SVG paths for each feature
 * @param features - Array of features to prepare
 * @param path - D3 path generator
 * @returns Array of prepared features with SVG paths
 */
function prepareFeatures(features, path) {
    if (!features || features.length === 0) {
        return [];
    }
    return features
        .map((feature) => {
        const svgPath = path(feature);
        if (!svgPath) {
            return null;
        }
        return {
            ...feature,
            svgPath,
        };
    })
        .filter((feature) => feature !== null);
}
/**
 * Creates a connector path between two coordinates
 * @param start - Starting coordinates [longitude, latitude]
 * @param end - Ending coordinates [longitude, latitude]
 * @param curve - D3 curve function for path interpolation
 * @returns SVG path string
 */
function createConnectorPath(start, end, curve) {
    // Type guard for curve function
    if (typeof curve !== 'function') {
        return '';
    }
    try {
        // Use type assertion for D3 curve - this is a known external API
        const curveFactory = curve;
        const line = curveFactory()
            .x((d) => d[0])
            .y((d) => d[1]);
        return line([start, end]) || '';
    }
    catch {
        return '';
    }
}

// Re-export utilities from focused modules
// Type guards and validation utilities remain in this file
// Advanced type guards for runtime type checking
// Geography data type guards
function isTopology(value) {
    if (typeof value !== 'object' || value === null)
        return false;
    const obj = value;
    return (obj.type === 'Topology' &&
        typeof obj.objects === 'object' &&
        obj.objects !== null &&
        Array.isArray(obj.arcs));
}
function isFeatureCollection(value) {
    if (typeof value !== 'object' || value === null)
        return false;
    const obj = value;
    return obj.type === 'FeatureCollection' && Array.isArray(obj.features);
}
function isFeature(value) {
    if (typeof value !== 'object' || value === null)
        return false;
    const obj = value;
    return obj.type === 'Feature' && 'geometry' in obj && 'properties' in obj;
}
function isValidGeometry(value) {
    if (typeof value !== 'object' || value === null)
        return false;
    const obj = value;
    if (!('type' in obj))
        return false;
    const validTypes = [
        'Point',
        'LineString',
        'Polygon',
        'MultiPoint',
        'MultiLineString',
        'MultiPolygon',
        'GeometryCollection',
    ];
    return validTypes.includes(obj.type);
}
// Coordinate type guards
function isValidLongitude(value) {
    return typeof value === 'number' && value >= -180 && value <= 180;
}
function isValidLatitude(value) {
    return typeof value === 'number' && value >= -90 && value <= 90;
}
function isValidCoordinates(value) {
    return (Array.isArray(value) &&
        value.length === 2 &&
        isValidLongitude(value[0]) &&
        isValidLatitude(value[1]));
}
// Projection type guards
function isGeoProjection(value) {
    return (typeof value === 'function' &&
        'invert' in value &&
        typeof value.invert === 'function');
}
function isProjectionName(value) {
    return (typeof value === 'string' && value.startsWith('geo') && value.length > 3);
}
// Error type guards
function isGeographyError(error) {
    if (!(error instanceof Error))
        return false;
    const errorObj = error;
    return ('type' in errorObj &&
        typeof errorObj.type === 'string' &&
        [
            'GEOGRAPHY_LOAD_ERROR',
            'GEOGRAPHY_PARSE_ERROR',
            'PROJECTION_ERROR',
            'VALIDATION_ERROR',
            'SECURITY_ERROR',
            'CONFIGURATION_ERROR',
            'CONTEXT_ERROR',
        ].includes(errorObj.type));
}
// URL validation type guard
function isValidGeographyUrl(value) {
    if (typeof value !== 'string')
        return false;
    try {
        const url = new URL(value);
        // Allow HTTPS and HTTP for localhost only
        if (url.protocol === 'https:')
            return true;
        if (url.protocol === 'http:' && url.hostname === 'localhost')
            return true;
        return false;
    }
    catch {
        return false;
    }
}
// Complex validation type guards
function isValidGeographyData(value) {
    return isTopology(value) || isFeatureCollection(value);
}
function isValidMapDimensions(width, height) {
    return (typeof width === 'number' &&
        typeof height === 'number' &&
        width > 0 &&
        height > 0 &&
        Number.isFinite(width) &&
        Number.isFinite(height));
}
// Factory function for creating custom type guards
function createTypeGuard(predicate) {
    return (value) => predicate(value);
}
// Utility functions to create enhanced geography errors
function createGeographyError(type, message, geography, details) {
    const error = new Error(message);
    error.type = type;
    if (geography)
        error.geography = geography;
    if (details)
        error.details = details;
    return error;
}
// Convenience functions for creating specific error types
function createValidationError(message, geography, details) {
    return createGeographyError('VALIDATION_ERROR', message, geography, details);
}
function createSecurityError(message, geography, details) {
    return createGeographyError('SECURITY_ERROR', message, geography, details);
}
function createProjectionError(message, geography, details) {
    return createGeographyError('PROJECTION_ERROR', message, geography, details);
}
function createConfigurationError(message, geography, details) {
    return createGeographyError('CONFIGURATION_ERROR', message, geography, details);
}
function createContextError(message, geography, details) {
    return createGeographyError('CONTEXT_ERROR', message, geography, details);
}

// React 19 resource preloading utilities for geography data
// Track preloaded URLs to avoid duplicate preloads
const preloadedUrls = new Set();
/**
 * Preload geography resources for better performance
 * Only preloads if the resource will be used soon
 */
function preloadGeography(url, immediate = false) {
    if (typeof url !== 'string' || !url) {
        return;
    }
    // Avoid duplicate preloads
    if (preloadedUrls.has(url)) {
        return;
    }
    try {
        const parsedUrl = new URL(url);
        // Always prefetch DNS and preconnect (lightweight operations)
        prefetchDNS(parsedUrl.origin);
        preconnect(parsedUrl.origin);
        // Only preload the actual resource if immediate or in production
        const shouldPreloadResource = immediate ||
            (typeof process !== 'undefined' && process.env.NODE_ENV === 'production');
        if (shouldPreloadResource) {
            preload(url, {
                as: 'fetch',
                crossOrigin: 'anonymous', // Most geography APIs support CORS
            });
            preloadedUrls.add(url);
        }
    }
    catch (error) {
        // Silently handle invalid URLs in development only
        if (typeof process !== 'undefined' &&
            process.env.NODE_ENV !== 'production') {
            // eslint-disable-next-line no-console
            console.warn('Failed to preload geography resource:', error);
        }
    }
}

var preloading = /*#__PURE__*/Object.freeze({
    __proto__: null,
    preloadGeography: preloadGeography
});

export { DEFAULT_GEOGRAPHY_FETCH_CONFIG, DEFAULT_SRI_CONFIG, DEVELOPMENT_GEOGRAPHY_FETCH_CONFIG, KNOWN_GEOGRAPHY_SRI, addCustomSRI, configureGeographySecurity, configureSRI, createConfigurationError, createConnectorPath, createContextError, createGeographyError, createGeographyFetchError, createProjectionError, createSecurityError, createTypeGuard, createValidationError, disableSRI, enableDevelopmentMode, enableStrictSRI, fetchGeographies, fetchGeographiesCache, generateSRIForUrls, generateSRIHash, getCoords, getFeatures, getMesh, getSRIForUrl, isFeature, isFeatureCollection, isGeoProjection, isGeographyError, isProjectionName, isString, isTopology, isValidCoordinates, isValidGeographyData, isValidGeographyUrl, isValidGeometry, isValidLatitude, isValidLongitude, isValidMapDimensions, preloadGeography$1 as preloadGeography, prepareFeatures, prepareMesh, readResponseWithSizeLimit, validateContentType, validateGeographyData, validateGeographyUrl, validateResponseSize, validateSRI };
